pkgs/development/python-modules/soundcard/default.nix

```
@@ -0,0 +1,49 @@
+{
+  buildPythonPackage,
+  cffi,
+  fetchFromGitHub,
+  lib,
+  libpulseaudio,
+  numpy,
+  setuptools,
+  testers,
+}:
+
+buildPythonPackage rec {
+  pname = "soundcard";
+  version = "0.4.3";
+  pyproject = true;
+
+  src = fetchFromGitHub {
+    owner = "bastibe";
+    repo = pname;
+    rev = version;
+    hash = "sha256-d91fhgkjnRFamMIn7hmXuHZt6kGBysAhQq8IsZ/WqHs=";
+  };
+
+  patchPhase = ''
+    substituteInPlace soundcard/pulseaudio.py \
+      --replace "'pulse'" "'${libpulseaudio}/lib/libpulse.so'"
+  '';
```

The newer `--replace-fail` flags should be used instead of the legacy `--replace` flag, since it won't silently fail.

Also, I think it's safer to also include `dlopen` in the replacement to not accidentally replace other instances of the `'pulse'` literal.

And finally, `libpulse.so` is only valid on linux systems. On darwin/Mac it's not `.so` but `.dylib`. `stdenv.hostPlatform.extensions.sharedLibrary` can be used instead, which evaluates to `.so` or `.dylib` depending on which system you're building for.


```suggestion
  patchPhase =
    let
      libpulse = "${libpulseaudio}/lib/libpulse${stdenv.hostPlatform.extensions.sharedLibrary}";
    in
    ''
      substituteInPlace soundcard/pulseaudio.py \
          --replace-fail "_ffi.dlopen('pulse')" "_ffi.dlopen('${libpulse}')"
    '';
```

---

Thank you very much for your advice here.
I don't have a MacOS to test right now, so I'm super grateful for feedback.

The [Upstream project](https://github.com/bastibe/SoundCard) says.

> SoundCard is a library for playing and recording audio without resorting to a CPython extension. Instead, it is implemented using the wonderful [CFFI](http://cffi.readthedocs.io/en/latest/) and the native audio libraries of Linux, Windows and macOS.

> SoundCard is cross-platform, and supports Linux/pulseaudio, Mac/coreaudio, and Windows/WASAPI. 


I'm not sure whether `pulseaudio` or `coraudio` should be used on a macOS with nix as package manager.

---

I'll try to look into it and try it out

---

Okay, I asked around on the Matrix chat about `dlopen` on darwin, still waiting for a response.

The current patching should be changed, then, to only happen on linux.
This can be done with
```nix
postPatch = lib.optionalString stdenv.isLinux ''
  substituteInPlace ....
'';
```

---

I just pushed a variant with `replace-fail`, where the patch is only applied (in `postPatch`) when we are on a linux system.
In this case I assume it's fine to use `.so` as extension.